File: nir/consensus_codec.py
"""Normative NIR consensus byte encoding shared by independent implementations."""

from __future__ import annotations

from hashlib import sha3_256
import re
import struct
import unicodedata


ENCODING_VERSION = 1
_PREFIX = b"NIR-CONSENSUS"
_DOMAIN = re.compile(r"^[A-Z0-9_-]{1,40}$")
_MAX_SAFE_INTEGER = 9_007_199_254_740_991
_MAX_DEPTH = 64
_MAX_CONTAINER_ENTRIES = 100_000
_MAX_STRING_BYTES = 16 * 1024 * 1024
_MAX_ENCODED_BYTES = 64 * 1024 * 1024


class ConsensusEncodingError(ValueError):
    """Raised when a value is outside the normative consensus data model."""


def _string_bytes(value: str, field: str) -> bytes:
    try:
        encoded = value.encode("utf-8", "strict")
    except UnicodeEncodeError as error:
        raise ConsensusEncodingError(
            f"{field} contains an unpaired Unicode surrogate"
        ) from error
    if len(encoded) > _MAX_STRING_BYTES:
        raise ConsensusEncodingError(f"{field} is too large")
    return encoded


def _u32(value: int) -> bytes:
    return struct.pack(">I", value)


def _append(output: bytearray, *parts: bytes) -> None:
    for part in parts:
        if len(output) + len(part) > _MAX_ENCODED_BYTES:
            raise ConsensusEncodingError("consensus value is too large")
        output.extend(part)


def _encode(
    value: object, output: bytearray, active: set[int], depth: int, field: str
) -> None:
    if depth > _MAX_DEPTH:
        raise ConsensusEncodingError("consensus value nesting is too deep")
    if value is None:
        _append(output, b"\x00")
        return
    if value is False:
        _append(output, b"\x01")
        return
    if value is True:
        _append(output, b"\x02")
        return
    if type(value) is int:
        if value < -_MAX_SAFE_INTEGER or value > _MAX_SAFE_INTEGER:
            raise ConsensusEncodingError(f"{field} must be an unambiguous safe integer")
        _append(output, b"\x03", struct.pack(">q", value))
        return
    if type(value) is str:
        encoded = _string_bytes(value, field)
        _append(output, b"\x04", _u32(len(encoded)), encoded)
        return
    if type(value) not in (list, dict):
        raise ConsensusEncodingError(f"{field} contains an unsupported value type")
    identity = id(value)
    if identity in active:
        raise ConsensusEncodingError("consensus value contains a cycle")
    active.add(identity)
    try:
        if type(value) is list:
            if len(value) > _MAX_CONTAINER_ENTRIES:
                raise ConsensusEncodingError(f"{field} is too large")
            _append(output, b"\x05", _u32(len(value)))
            for index, item in enumerate(value):
                _encode(item, output, active, depth + 1, f"{field}[{index}]")
            return
        if len(value) > _MAX_CONTAINER_ENTRIES:
            raise ConsensusEncodingError(f"{field} has too many fields")
        entries: list[tuple[bytes, str, object]] = []
        key_bytes_total = 0
        for key, item in value.items():
            if type(key) is not str:
                raise ConsensusEncodingError(f"{field} contains a non-string key")
            key_bytes = _string_bytes(key, f"{field} key")
            key_bytes_total += len(key_bytes)
            if key_bytes_total > _MAX_ENCODED_BYTES:
                raise ConsensusEncodingError("consensus value is too large")
            if unicodedata.normalize("NFC", key) != key:
                raise ConsensusEncodingError(f"{field} contains an ambiguous non-NFC key")
            entries.append((key_bytes, key, item))
        entries.sort(key=lambda entry: entry[0])
        _append(output, b"\x06", _u32(len(entries)))
        for key_bytes, key, item in entries:
            _append(output, b"\x04", _u32(len(key_bytes)), key_bytes)
            _encode(item, output, active, depth + 1, f"{field}.{key}")
    finally:
        active.remove(identity)


def consensus_value_bytes(value: object) -> bytes:
    encoded = bytearray()
    _encode(value, encoded, set(), 0, "consensus value")
    return bytes(encoded)


def consensus_envelope_bytes(domain: str, value: object) -> bytes:
    if type(domain) is not str or _DOMAIN.fullmatch(domain) is None:
        raise ConsensusEncodingError("invalid cryptographic domain")
    domain_bytes = domain.encode("ascii")
    return (
        _PREFIX + b"\x00" + struct.pack(">H", ENCODING_VERSION) +
        bytes((len(domain_bytes),)) + domain_bytes + consensus_value_bytes(value)
    )


def consensus_hash(domain: str, value: object) -> str:
    return sha3_256(consensus_envelope_bytes(domain, value)).hexdigest()

File: blockchain/consensus-codec.mjs
import { types as utilTypes } from "node:util";

export const CONSENSUS_ENCODING_VERSION = 1;

const PROTOCOL_ENCODINGS = Object.freeze(new Map([
  [24, 1],
  [25, 1],
  [26, 1],
  [27, 1],
  [28, 1],
  [29, 1],
  [30, 1],
  [31, 1],
  [32, 1],
]));

const PREFIX = Buffer.from("NIR-CONSENSUS", "ascii");
const MAX_DEPTH = 64;
const MAX_CONTAINER_ENTRIES = 100_000;
const MAX_STRING_BYTES = 16 * 1024 * 1024;
const MAX_ENCODED_BYTES = 64 * 1024 * 1024;
const DOMAIN = /^[A-Z0-9_-]{1,40}$/;

export function consensusEncodingVersionForProtocol(protocolVersion) {
  const version = PROTOCOL_ENCODINGS.get(protocolVersion);
  if (version === undefined) {
    throw new Error(`protocol version ${protocolVersion} has no consensus encoding`);
  }
  return version;
}

function u32(value) {
  const encoded = Buffer.allocUnsafe(4);
  encoded.writeUInt32BE(value);
  return encoded;
}

function validUnicodeScalarString(value) {
  for (let index = 0; index < value.length; index += 1) {
    const unit = value.charCodeAt(index);
    if (unit >= 0xd800 && unit <= 0xdbff) {
      const following = value.charCodeAt(index + 1);
      if (!(following >= 0xdc00 && following <= 0xdfff)) return false;
      index += 1;
    } else if (unit >= 0xdc00 && unit <= 0xdfff) {
      return false;
    }
  }
  return true;
}

function stringBytes(value, field) {
  if (!validUnicodeScalarString(value)) {
    throw new Error(`${field} contains an unpaired Unicode surrogate`);
  }
  const encoded = Buffer.from(value, "utf8");
  if (encoded.length > MAX_STRING_BYTES) throw new Error(`${field} is too large`);
  return encoded;
}

function ordinaryPrototype(value, expected, field) {
  if (utilTypes.isProxy(value) || Object.getPrototypeOf(value) !== expected) {
    throw new Error(`${field} has an exotic prototype or proxy`);
  }
}

function dataDescriptor(value, key, field) {
  const descriptor = Object.getOwnPropertyDescriptor(value, key);
  if (!descriptor || !("value" in descriptor) || !descriptor.enumerable) {
    throw new Error(`${field} contains an accessor or non-enumerable field`);
  }
  return descriptor.value;
}

function sortedObjectEntries(value, field) {
  ordinaryPrototype(value, Object.prototype, field);
  const keys = Reflect.ownKeys(value);
  if (keys.length > MAX_CONTAINER_ENTRIES || keys.some((key) => typeof key !== "string")) {
    throw new Error(`${field} has too many fields or a symbol key`);
  }
  const normalized = new Set();
  let keyBytes = 0;
  const entries = keys.map((key) => {
    const bytes = stringBytes(key, `${field} key`);
    keyBytes += bytes.length;
    if (keyBytes > MAX_ENCODED_BYTES) throw new Error("consensus value is too large");
    const nfc = key.normalize("NFC");
    if (nfc !== key || normalized.has(nfc)) {
      throw new Error(`${field} contains an ambiguous non-NFC key`);
    }
    normalized.add(nfc);
    return { bytes, key, value: dataDescriptor(value, key, field) };
  });
  entries.sort((left, right) => Buffer.compare(left.bytes, right.bytes));
  return entries;
}

function arrayValues(value, field) {
  ordinaryPrototype(value, Array.prototype, field);
  if (value.length > MAX_CONTAINER_ENTRIES) throw new Error(`${field} is too large`);
  const keys = Reflect.ownKeys(value);
  if (keys.some((key) => typeof key !== "string" ||
      (key !== "length" && !/^(0|[1-9][0-9]*)$/.test(key)))) {
    throw new Error(`${field} contains a symbol or extra field`);
  }
  if (keys.length !== value.length + 1) throw new Error(`${field} is sparse`);
  const values = [];
  for (let index = 0; index < value.length; index += 1) {
    const key = String(index);
    if (!Object.hasOwn(value, key)) throw new Error(`${field} is sparse`);
    values.push(dataDescriptor(value, key, field));
  }
  return values;
}

function append(context, ...values) {
  for (const value of values) {
    if (context.total + value.length > MAX_ENCODED_BYTES) {
      throw new Error("consensus value is too large");
    }
    context.total += value.length;
    context.chunks.push(value);
  }
}

function encodeValue(value, context, active, depth, field) {
  if (depth > MAX_DEPTH) throw new Error("consensus value nesting is too deep");
  if (value === null) {
    append(context, Buffer.from([0x00]));
    return;
  }
  if (value === false) {
    append(context, Buffer.from([0x01]));
    return;
  }
  if (value === true) {
    append(context, Buffer.from([0x02]));
    return;
  }
  if (typeof value === "number") {
    if (!Number.isSafeInteger(value) || Object.is(value, -0)) {
      throw new Error(`${field} must be an unambiguous safe integer`);
    }
    const encoded = Buffer.allocUnsafe(9);
    encoded[0] = 0x03;
    encoded.writeBigInt64BE(BigInt(value), 1);
    append(context, encoded);
    return;
  }
  if (typeof value === "string") {
    const encoded = stringBytes(value, field);
    append(context, Buffer.from([0x04]), u32(encoded.length), encoded);
    return;
  }
  if (typeof value !== "object") {
    throw new Error(`${field} contains an unsupported value type`);
  }
  if (active.has(value)) throw new Error("consensus value contains a cycle");
  active.add(value);
  try {
    if (Array.isArray(value)) {
      const values = arrayValues(value, field);
      append(context, Buffer.from([0x05]), u32(values.length));
      values.forEach((entry, index) => {
        encodeValue(entry, context, active, depth + 1, `${field}[${index}]`);
      });
      return;
    }
    const entries = sortedObjectEntries(value, field);
    append(context, Buffer.from([0x06]), u32(entries.length));
    for (const entry of entries) {
      append(context, Buffer.from([0x04]), u32(entry.bytes.length), entry.bytes);
      encodeValue(entry.value, context, active, depth + 1, `${field}.${entry.key}`);
    }
  } finally {
    active.delete(value);
  }
}

export function consensusValueBytes(value) {
  const context = { chunks: [], total: 0 };
  encodeValue(value, context, new WeakSet(), 0, "consensus value");
  return Buffer.concat(context.chunks, context.total);
}

function envelopePrefix(domain, encodingVersion) {
  if (!DOMAIN.test(domain ?? "")) throw new Error("invalid cryptographic domain");
  if (encodingVersion !== CONSENSUS_ENCODING_VERSION) {
    throw new Error("unsupported consensus encoding version");
  }
  const domainBytes = Buffer.from(domain, "ascii");
  return Buffer.concat([
    PREFIX,
    Buffer.from([0x00, 0x00, encodingVersion, domainBytes.length]),
    domainBytes,
  ]);
}

export function consensusArrayEnvelopeHeader(domain, length, {
  encodingVersion = CONSENSUS_ENCODING_VERSION,
} = {}) {
  if (!Number.isSafeInteger(length) || length < 0 || length > MAX_CONTAINER_ENTRIES) {
    throw new Error("consensus array length is invalid");
  }
  return Buffer.concat([
    envelopePrefix(domain, encodingVersion), Buffer.from([0x05]), u32(length),
  ]);
}

export function consensusEnvelopeBytes(domain, value, options = {}) {
  return Buffer.concat([
    envelopePrefix(domain, options.encodingVersion ?? CONSENSUS_ENCODING_VERSION),
    consensusValueBytes(value),
  ]);
}

function jsonValue(value, active, depth, field) {
  if (depth > MAX_DEPTH) throw new Error("consensus value nesting is too deep");
  if (value === null || typeof value === "boolean") return JSON.stringify(value);
  if (typeof value === "number") {
    if (!Number.isSafeInteger(value) || Object.is(value, -0)) {
      throw new Error(`${field} must be an unambiguous safe integer`);
    }
    return String(value);
  }
  if (typeof value === "string") {
    stringBytes(value, field);
    return JSON.stringify(value);
  }
  if (typeof value !== "object") {
    throw new Error(`${field} contains an unsupported value type`);
  }
  if (active.has(value)) throw new Error("consensus value contains a cycle");
  active.add(value);
  try {
    if (Array.isArray(value)) {
      const values = arrayValues(value, field);
      return `[${values.map((entry, index) =>
        jsonValue(entry, active, depth + 1, `${field}[${index}]`)).join(",")}]`;
    }
    const entries = sortedObjectEntries(value, field);
    return `{${entries.map((entry) => `${JSON.stringify(entry.key)}:${
      jsonValue(entry.value, active, depth + 1, `${field}.${entry.key}`)}`).join(",")}}`;
  } finally {
    active.delete(value);
  }
}

export function strictCanonicalJson(value) {
  // Enforce the same aggregate bounds before constructing a JSON string.
  consensusValueBytes(value);
  return jsonValue(value, new WeakSet(), 0, "consensus value");
}
